### DISCUSSION TITLE
Type of property 'client' circularly references itself in mapped type 'EntityDTO<Loaded<Event<?>, any>>'

### DISCUSSION THREAD

**Discussion by micchickenburger** [mikro-orm/mikro-orm#3511] [Q&A] [answered]
**Describe the bug**

After updating from mikro-orm from v5.2.4 to v5.4.2 I'm getting a strange Typescript compilation error:

```
Type of property 'events' circularly references itself in mapped type 'EntityDTO<Client & { name: string; events: Collection<Event<Client>, object> & LoadedCollection<Loaded<Event<Client>, never>>; ... 13 more ...; setSchema: ((schema?: string | undefined) => void) & {}; }>'.     ts(2615)
```

I have class Event that has a ManyToOne relationship with Client on `Client.events`.

```typescript
@Entity()
class Event<T> extends BaseEntity<Event<T>> {
  // ...

  @Property({ persist: false })
  get entity() {
    return this.client || this.user; // error occurs here
  }

  @ManyToOne()
    client?: Client;

  @ManyToOne()
    user?: User;

  // ...
}

@Entity()
class Client extends BaseEntity<Client> {
  // ...

  @OneToMany(() => Event<Client>, (event) => event.client)
    events = new Collection<Event<Client>>(this);

  // ...
}
```

**Expected behavior**
This same code worked before updating, so I wonder if some recent changes on types have caused Typescript to detect this circular dependency as a problem.  Expected behavior was as prior.

**Versions**

| Dependency | Version |
| - | - |
| node | v16.16.0 |
| typescript | v4.7.4 & 4.8.3 |
| mikro-orm | 5.4.2 |
| postgresql driver | 5.4.2 |

**Comment by B4nan** [comment, marked as answer]
I would guess this is because you upgraded TypeScript. There is nothing to fix at our end, the error is coming from your code, and I believe it is just saying that you need to provide explicit return type because it is not possible to infer it (due to the circular reference).

**Reply by micchickenburger** [reply]
Thanks for your reply.  I didn't upgrade TypeScript between Mikro-ORM updates, only once the issue occurred to see if it would persist across TypeScript versions.  But you are right that specifying the return type solves the problem:

```typescript
  @Property({ persist: false })
  get entity(): Client | User {
    return this.client as Client || this.user as User;
  }
```
